crates/bevy_gilrs/src/rumble.rs

```diff
@@ -0,0 +1,146 @@
+) -> Vec<ff::BaseEffect> {
+    let mut effects = Vec::new();
+    if strong > 0. {
+        effects.push(BaseEffect {
+            kind: BaseEffectType::Strong {
+                magnitude: to_gilrs_magnitude(strong),
+            },
+            ..Default::default()
+        });
+    }
+    if weak > 0. {
+        effects.push(BaseEffect {
+            kind: BaseEffectType::Strong {
+                magnitude: to_gilrs_magnitude(weak),
+            },
+            ..Default::default()
+        });
+    }
+    effects
+}
+
+fn handle_rumble_request(
+    manager: &mut RumblesManager,
+    gilrs: &mut Gilrs,
+    rumble: GamepadRumbleRequest,
+    current_time: f32,
+) -> Result<(), RumbleError> {
+    let gamepad = match rumble {
+        GamepadRumbleRequest::Add { gamepad, .. } | GamepadRumbleRequest::Stop { gamepad } => {
+            gamepad
+        }
+    };
+    let (gamepad_id, _) = gilrs
+        .gamepads()
+        .find(|(pad_id, _)| convert_gamepad_id(*pad_id) == gamepad)
+        .ok_or(RumbleError::GamepadNotFound)?;
+
+    match rumble {
+        GamepadRumbleRequest::Stop { .. } => {
+            // `ff::Effect` uses RAII, dropping = deactivating
+            manager.rumbles.remove(&gamepad_id);
+        }
+        GamepadRumbleRequest::Add {
+            duration,
+            intensity,
+            ..
+        } => {
+            let deadline = current_time + duration.as_secs_f32();
+
+            let mut effect_builder = ff::EffectBuilder::new();
+
+            for effect in get_base_effects(intensity) {
+                effect_builder.add_effect(effect);
+            }
+
+            let effect = effect_builder.gamepads(&[gamepad_id]).finish(gilrs)?;
+            effect.play()?;
+
+            let gamepad_rumbles = manager.rumbles.entry(gamepad_id).or_default();
+            gamepad_rumbles.push(RunningRumble { deadline, effect });
+        }
+    }
+
+    Ok(())
+}
+pub(crate) fn play_gilrs_rumble(
+    time: Res<Time>,
+    mut gilrs: NonSendMut<Gilrs>,
+    mut requests: EventReader<GamepadRumbleRequest>,
+    mut manager: NonSendMut<RumblesManager>,
+) {
+    let current_time = time.elapsed_seconds();
+    // Remove outdated rumble effects.
+    for (_gamepad, rumbles) in manager.rumbles.iter_mut() {
+        // `ff::Effect` uses RAII, dropping = deactivating
+        rumbles.retain(|RunningRumble { deadline, .. }| *deadline >= current_time);
+    }
+    manager
+        .rumbles
+        .retain(|_gamepad, rumbles| !rumbles.is_empty());
+
+    // Add new effects.
+    for rumble in requests.iter().cloned() {
+        let gamepad = match rumble {
+            GamepadRumbleRequest::Add { gamepad, .. } | GamepadRumbleRequest::Stop { gamepad } => {
+                gamepad
+            }
+        };
+        match handle_rumble_request(&mut manager, &mut gilrs, rumble, current_time) {
+            Ok(()) => {}
+            Err(RumbleError::GilrsError(err)) => {
+                debug!(
```

Comment:
```suggestion
                warn!(
```

I think these errors are similar severity, and `debug` warnings are hidden by default.

Reply:
I left force feedback not supported as a debug, since we currently don't have a way to query for force feedback support for a controller (avoid spamming warnings the user can't solve)